### ISSUE TITLE
division with Int datatype is incorrect in JavaScript generated code

### ISSUE THREAD

**Issue by KesterTong** [idris-lang/Idris-dev#1818]
The following test case illustrates the problem:
```
tenDivTwelve : Int
tenDivTwelve = 10 `div` 12

isTenDivTwelveLeqZero : Bool
isTenDivTwelveLeqZero = if tenDivTwelve <= 0 then True else False

main : IO ()
main = do
  putStrLn $ "10 / 12 = " ++ (show tenDivTwelve)
  putStrLn $ "10 / 12 <= 0 is " ++ (show isTenDivTwelveLeqZero)
```

When compiled to machine code this results in the output
```
10 / 12 = 0
10 / 12 <= 0 is True
```

When compiled to JavaScript this results in the (console) output
```
10 / 12 = 0.8333333333333334
10 / 12 <= 0 is False
```

It is possible that this bug is as simple as replacing ```x / y``` with ```Math.floor(x / y)``` inside the JavaScript code generated for the ``` `div` ``` function.

**Comment by ahmadsalim** [comment]
Still seems to be an issue.

One could fix this by compiling division of integers to:
`~~(~~10 / ~~12)`

**Comment by KesterTong** [comment]
I agree that `~~` is better than my suggestion of `Math.floor`.  Asm.js is a useful reference since it is implemented in pure JS and has integer types.  I don't know enough about JS minutiae to know why asm.js uses both `~~` and `|0`.

However I don't think coercing the inputs is necessary, since they are guaranteed to be the correct type, assuming every other function works correctly.

It might help to take a look at asm.js and use the same model for the FFI, (e.g. iirc, `Int` is implementation defined so it would make sense for `Int` to be the asm.js int type, which is an int32).  I don't have much time to work on this myself, and it probably isn't a priority for Idris.  I think `~~(10/12)` would be a good enough solution.

**Comment by ahmadsalim** [comment]
@KesterTong Thanks, I think it could definitely be work looking into asm.js.
